dipy/denoise/tests/test_ascm.py

```diff
@@ -0,0 +1,111 @@
+    assert_array_almost_equal(S0, S0n)
+
+
+def test_ascm_random_noise():
+    S0 = 100 + 2 * np.random.standard_normal((22, 23, 30))
+    S0n1 = nlmeans(S0, sigma=np.ones((22, 23, 30)), rician=False,
+                   patch_radius=1, block_radius=1, avg_type='blockwise')
+    S0n2 = nlmeans(S0, sigma=np.ones((22, 23, 30)), rician=False,
+                   patch_radius=2, block_radius=1, avg_type='blockwise')
+    S0n = ascm(S0, S0n1, S0n2, 1)
+
+    print(S0.mean(), S0.min(), S0.max())
+    print(S0n.mean(), S0n.min(), S0n.max())
+
+    assert_(S0n.min() > S0.min())
+    assert_(S0n.max() < S0.max())
+    assert_equal(np.round(S0n.mean()), 100)
+
+
+def test_ascm_rmse_with_nlmeans():
+    # checks the smoothness
+    S0 = np.ones((30, 30, 30)) * 100
+    S0[10:20, 10:20, 10:20] = 50
+    S0[20:30, 20:30, 20:30] = 0
+    S0_noise = S0 + 20 * np.random.standard_normal((30, 30, 30))
+    print("Original RMSE", np.sum(np.abs(S0 - S0_noise)) / np.sum(S0))
+
+    S0n1 = nlmeans(
+        S0_noise,
+        sigma=400,
+        rician=False,
+        patch_radius=1,
+        block_radius=1,
+        avg_type='blockwise')
+    print("Smaller patch RMSE", np.sum(np.abs(S0 - S0n1)) / np.sum(S0))
+    S0n2 = nlmeans(
+        S0_noise,
+        sigma=400,
+        rician=False,
+        patch_radius=2,
+        block_radius=2,
+        avg_type='blockwise')
+    print("Larger patch RMSE", np.sum(np.abs(S0 - S0n2)) / np.sum(S0))
+    S0n = ascm(S0, S0n1, S0n2, 400)
+    print("ASCM RMSE", np.sum(np.abs(S0 - S0n)) / np.sum(S0))
+
+    assert_(np.sum(np.abs(S0 - S0n)) / np.sum(S0) <
+            np.sum(np.abs(S0 - S0n1)) / np.sum(S0))
+    assert_(
+        np.sum(
+            np.abs(
+                S0 -
+                S0n)) /
+        np.sum(S0) < np.sum(
+            np.abs(
+                S0 -
+                S0_noise)) /
+        np.sum(S0))
+    assert_(90 < np.mean(S0n) < 110)
+
+
+def test_sharpness():
+    # check the edge-preserving nature
+    S0 = np.ones((30, 30, 30)) * 100
+    S0[10:20, 10:20, 10:20] = 50
+    S0[20:30, 20:30, 20:30] = 0
+    S0_noise = S0 + 20 * np.random.standard_normal((30, 30, 30))
+    S0n1 = nlmeans(
+        S0_noise,
+        sigma=400,
+        rician=False,
+        patch_radius=1,
+        block_radius=1,
+        avg_type='blockwise')
+    edg1 = np.abs(np.mean(S0n1[8, 10:20, 10:20] - S0n1[12, 10:20, 10:20]) - 50)
+    print("Edge gradient smaller patch", edg1)
+    S0n2 = nlmeans(
+        S0_noise,
+        sigma=400,
+        rician=False,
+        patch_radius=2,
+        block_radius=2,
+        avg_type='blockwise')
+    edg2 = np.abs(np.mean(S0n2[8, 10:20, 10:20] - S0n2[12, 10:20, 10:20]) - 50)
+    print("Edge gradient larger patch", edg2)
+    S0n = ascm(S0, S0n1, S0n2, 400)
+    edg = np.abs(np.mean(S0n[8, 10:20, 10:20] - S0n[12, 10:20, 10:20]) - 50)
+    print("Edge gradient ASCM", edg)
+
+    assert_(edg2 > edg1)
+    assert_(edg2 > edg)
+    assert_(np.abs(edg1 - edg) < 1.5)
+
```

Comment:
Add test where you denoise the same volume with your ascm and with Omar's ascm and you measure the error of the final volumes. You will run Omar's ascm externally and just save the denoised result in dipy.data. Use a small volume e.g. look for an existing T1 if any. We need to keep data in dipy/data of a small size.